fsthttp/request.go

```diff
@@ -399,40 +423,308 @@ func (req *Request) Send(ctx context.Context, backend string) (*Response, error)
+	abiResp, abiBody, err := req.sendWithoutCaching(backend)
 	if err != nil {
-		return nil, fmt.Errorf("construct response: %w", err)
+		return nil, err
 	}
 
+	resp, err = newResponse(req, backend, abiResp, abiBody)
+	if err != nil {
+		return nil, err
+	}
+	resp.updateFastlyCacheHeaders(req)
 	return resp, nil
+
+}
+
+func newRequestFromHandle(reqh *fastly.HTTPRequest, body io.ReadCloser, headers Header, options CacheOptions) (*Request, error) {
+	method, err := reqh.GetMethod()
+	if err != nil {
+		return nil, err
+	}
+	url, err := reqh.GetURI()
+	if err != nil {
+		return nil, err
+	}
+
+	req, _ := NewRequest(method, url, body)
+	req.CacheOptions = options
+	req.Header = headers
+	req.abi.req = reqh
+	req.abi.body, _ = abiBodyFrom(body)
+
+	return req, nil
+}
+
+type pendingBackendRequestForCaching struct {
+	cacheHandle  *fastly.HTTPCacheHandle
+	pending      *fastly.PendingRequest
+	afterSend    func(*CandidateResponse) error
+	cacheOptions CacheOptions
+	req          *Request
+}
+
+func (req *Request) sendAsyncForCaching(ctx context.Context, cacheHandle *fastly.HTTPCacheHandle, backend string) (*pendingBackendRequestForCaching, error) {
+
+	reqh, err := fastly.HTTPCacheGetSuggestedBackendRequest(cacheHandle)
+	if err != nil {
+		return nil, fmt.Errorf("get suggested backend request: %w", err)
+	}
+
+	suggReq, err := newRequestFromHandle(reqh, req.Body, req.Header, req.CacheOptions)
+	if err != nil {
+		return nil, err
+	}
+
+	if suggReq.CacheOptions.BeforeSend != nil {
+		if err := suggReq.CacheOptions.BeforeSend(suggReq); err != nil {
+			// TODO(dgryski): sentinel ErrReject ?
+			return nil, err
+		}
+	}
+
+	// If BeforeSend calls SetBody, abi.body will be updated for the new Body
+	finalCacheOptions := suggReq.CacheOptions
+	suggReq.setABIRequestOptions()
+
+	// copy body
+	var errc chan error = make(chan error, 3)
+	suggReq.copyBody(errc)
+
+	abiPending, err := suggReq.sendAsyncWithoutCaching(ctx, backend)
+	if err != nil {
+		return nil, err
+	}
+
+	return &pendingBackendRequestForCaching{
+		cacheHandle:  cacheHandle,
+		req:          suggReq,
+		pending:      abiPending,
+		afterSend:    req.CacheOptions.AfterSend,
+		cacheOptions: finalCacheOptions,
+	}, nil
+}
+
+func (req *Request) sendAsyncWithoutCaching(_ context.Context, backend string) (*fastly.PendingRequest, error) {
+	abiPending, err := req.abi.req.SendAsyncV2(req.abi.body, backend, false)
+
+	if err != nil {
+		return nil, fmt.Errorf("send async: %w", err)
+	}
+
+	return abiPending, nil
+}
+
+func (req *Request) sendWithoutCaching(backend string) (*fastly.HTTPResponse, *fastly.HTTPBody, error) {
+	abiResp, abiBody, err := req.abi.req.SendV3(req.abi.body, backend)
+	if err != nil {
+		return nil, nil, fmt.Errorf("send v3: %w", err)
+	}
+	return abiResp, abiBody, nil
+}
+
+func (req *Request) mustUseHostCaching() (bool, error) {
+	ok, err := fastly.HTTPCacheIsRequestCacheable(req.abi.req)
+	return !ok, err
+}
+
+func (req *Request) shouldUseGuestCache() (bool, error) {
```

Comment:
Should this be called `shouldUseGuestCaching` for consistency?

Reply:
4b4f353